Add Install Records for VS Managed Workloads During 'Update'

nagilson opened the pull request:
# Why are we doing this 
This is an attempt to fix the following: https://github.com/dotnet/sdk/issues/21811 and https://github.com/dotnet/aspire/issues/1559

If The .NET SDK CLI updates a workload, it will then update the manifests. VS will also begin using the updated manifests as it shares the workload resolver with the CLI. But Visual Studio will not have updated its workloads, and the workload resolver does not have logic to use the old manifests for VS.

Then, Visual Studio will reference to nonexistent packs because its workloads were not updated even though the manifest points to newer packs. So, by writing the install record, we hope to make the rest of the update command add back in the packs, so VS would work fine. And if we uninstall the standalone SDK, then the old packs still have a VS reference, and the newer manifest should be removed, so it should be fine.

# What does this PR do

Makes the CLI update command create SDK install records for VS managed workloads that don't already have SDK managed records. This is to cause the downstream logic to install the packs for VS to use.

# Tests

Lots of manual testing. Which requires many hours of setup, don't recommend.
This is hard to write tests for -- maybe once VM changes are merged we can add tests.
For evidence of behavior, see screenshots below.

# Before

![image](https://github.com/dotnet/sdk/assets/23152278/83f65ee6-211a-46bb-8987-fa919de55136)

Running `workload update` from this state does not create an install record on the SDK side.

Also, after running update:

`dotnet publish` now becomes broken for VS workload apps, such as wasm apps, or for aspire apps.

![image](https://github.com/dotnet/sdk/assets/23152278/4db1c830-7157-43ce-9ae6-7a35cda801b0)

Also, installing any other workload via CLI (in this example, MAUI), the VS workload(s) (wasm-tools in this case) are no longer found and you are in a broken state. Running workload install again does not fix the problem.

Without my fix and Phil's fix in Visual Studio, templates are also missing from VS Workloads after installing a different workload in the CLI.
![image](https://github.com/dotnet/sdk/assets/23152278/40292562-7175-4b03-9d28-1481651debb0)


# After 😮

When running `dotnet workload update`, install records are created for VS workloads in the CLI side.
![image](https://github.com/dotnet/sdk/assets/23152278/eb137993-fb77-4112-a185-a68d874cb09a)

After installing MAUI or another workload, wasm-tools is still found.
![image](https://github.com/dotnet/sdk/assets/23152278/4874cd9c-7fc6-4647-8a41-741c6274b5d7)

`dotnet publish` succeeds for apps with workloads that were previously installed by VS and got updated by the CLI.
![image](https://github.com/dotnet/sdk/assets/23152278/d666e4dd-d638-4a3b-b6fa-c51269beed82)

Templates appear in VS even after installing or updating workloads via the CLI.
<img width="298" alt="Screenshot_19" src="https://github.com/dotnet/sdk/assets/23152278/1c00be8d-4cc2-43c6-b817-0ca7f655c6dd">

When running install, a message appears indicating workloads from VS had install records added to them.
Example:
![image](https://github.com/dotnet/sdk/assets/23152278/0c92fbb2-f0d5-4be3-be47-21fc0c608ce4)

nagilson committed:
Try adding install records for VS only workloads during 'workload update'

nagilson committed:
Switch from List to IEnumerable because List is unnecessary

joeloff commented on src/Cli/dotnet/commands/dotnet-workload/update/WorkloadUpdateCommand.cs:
```diff
@@ -139,6 +144,32 @@ public void UpdateWorkloads(bool includePreviews = false, DirectoryPath? offline
+#if !DOT_NET_BUILD_FROM_SOURCE
+            if (OperatingSystem.IsWindows())
+            {
+                // Do this gross check to avoid adding an unused & unnecessary method to FileBasedInstallers
```
Please remove "gross" - this can trigger policheck issues.

Also, you can query whether we're an MSI or non-MSI based installation instead of looking at the actual type.

nagilson replied:
Good point, thank you. Is that query based on the MSI sentinel file? I actually think this may be the best choice for type safety. For example, if the code to query our installation type is out of sync with the type of `_workloadInstaller` due to some bug, then this would become an invalid type cast.

Probably I prefer to do this than to have an empty function in the `FileBasedInstaller`. But if there is a better alternative, I am certainly open to it. Type checking isn't all that awful in all cases imo, I am just hesitant to use it :)

nagilson committed:
Add VS Workload Install Record Creation for both Workload Install and Restore

nagilson committed:
Cast a workload installer, not a resolver (copy paste error)

nagilson commented:
/azp run

nagilson committed:
Merge remote-tracking branch 'upstream/release/8.0.3xx' into nagilson-vs-wkl-update

nagilson committed:
Move Record Creation to before installed workloads are queried so It executes properly and fixes preexisting bad states even if skip manifest update is on

nagilson committed:
Merge remote-tracking branch 'upstream/release/8.0.3xx' into nagilson-vs-wkl-update

nagilson committed:
Code Cleanup

Change reporter output to be localized
Reduce if statement logic and optimize code

nagilson committed:
Remove extraneous code that does extra work

The workloads with install records from the sdk are already found in the passed parameter so we dont need to filter them out twice

nagilson committed:
Simplify the output message for workload record creation

nagilson commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -148,13 +150,16 @@ public void InstallWorkloads(IEnumerable<WorkloadId> workloadIds, bool skipManif
                     Reporter.WriteLine(LocalizableStrings.CheckForUpdatedWorkloadManifests);
                 }
-                // Update currently installed workloads
+
```
```suggestion
```

nagilson committed:
Remove extraneous whitespace from removed code

nagilson commented on src/Cli/dotnet/commands/dotnet-workload/update/WorkloadUpdateCommand.cs:
```diff
@@ -119,7 +122,9 @@ public void UpdateWorkloads(bool includePreviews = false, DirectoryPath? offline
         {
             Reporter.WriteLine();
 
+
```
```suggestion
```

nagilson committed:
Remove extraneous whitespace from removed code

nagilson committed:
Condition the workloads to write records for message based on IFF there are any to write

joeloff reviewed (approved):
Looks good. Couple of scenario's you'll want to test once you have official builds of the sdk in installer.

1. Install VS + wasm tools
2. Install the SDK and run `dotnet workload update --skip-manifest-update` to add the CLI references
3. Uninstall VS
4. Uninstall the SDK
5. Verify that the SDK removes the WASM workload installations

Variation on this is to alter the later steps
3. Remove the SDK (this should remove the CLI dependents)
4. Remove VS
5. Verify that VS was able to remove WASM tools (because step 3 should have removed the dependents that would otherwise have blocked the removal)

Another variation is still install multiple SDKs after VS, different feature bands, like 8.0.3xx and eventually 8.0.4xx

nagilson commented:
@joeloff  Thank you for your review! Are those things we can test before getting an official build just via patching a standalone sdk? It seems like all of those I should be able to test before merging, except for the 8.0.3xx one but I might be wrong.

joeloff commented:
No, at the moment you need to get an official build once your SDK changes flow to installer. This is another good example where merging the repos has real benefits

nagilson commented:
Ok, thanks :) In that case, I suppose we will merge it!

dsplaisted commented:
> No, at the moment you need to get an official build once your SDK changes flow to installer. This is another good example where merging the repos has real benefits

I think you could test this without waiting for flow into installer by copying the stage 2 implementation over the installed SDK folder.  I showed Noah how to do this manually, and the VM installation tests automate this.

joeloff commented:
> > No, at the moment you need to get an official build once your SDK changes flow to installer. This is another good example where merging the repos has real benefits
> 
> I think you could test this without waiting for flow into installer by copying the stage 2 implementation over the installed SDK folder. I showed Noah how to do this manually, and the VM installation tests automate this.

It won't work for this specific case. You want to verify that running the bundle cleanly removes everything. Installing an MSI and then modifying the files won't let the MSI remove the files. They'll get flagged as user-modified and left behind. 

You could maybe push your branch to internal, grab the toolset zip, then modify the installer repo to use that zip and build a more official bundle.

nagilson commented:
It looks like it got in 2 hours after this PR, https://github.com/dotnet/installer/pull/19037 so the nightly build should have it. Let me try it out.

nagilson commented:
Doing the testing now. One thing I noticed is these messages side by side might confuse people but I think it's ok, since the 2nd message is prefixed with 'VS' workloads.
![image](https://github.com/dotnet/sdk/assets/23152278/3b2b3421-3148-4c49-9a2f-1d441b06f602)

nagilson commented:
> 1. Install VS + wasm tools
> 2. Install the SDK and run `dotnet workload update` to add the CLI references
> 3. Uninstall VS
> 4. Uninstall the SDK
> 5. Verify that the SDK removes the WASM workload installations

This works, the entire dotnet folder is gone.
![image](https://github.com/dotnet/sdk/assets/23152278/e7cdae5b-8abe-4fc7-84cd-16623cd299f1)

nagilson commented:
> Variation on this is to alter the later steps 3. Remove the SDK (this should remove the CLI dependents) 4. Remove VS 5. Verify that VS was able to remove WASM tools (because step 3 should have removed the dependents that would otherwise have blocked the removal)

This also works. Only dotnet.exe remains.
![image](https://github.com/dotnet/sdk/assets/23152278/5f36c87d-7e4c-4a52-8ddf-058990d80a13)


It also worked in between the state of after uninstalling the standalone SDK. The VS workloads were still in a good state. 
![image](https://github.com/dotnet/sdk/assets/23152278/18803c15-3444-484f-a2c1-1aa197b0874e)

nagilson commented:
> Another variation is still install multiple SDKs after VS, different feature bands, like 8.0.3xx and eventually 8.0.4xx

I tried using 8.0.202, mixing operations between 8.0.100 and 8.0.300, and VS, and pinning to it and doing various things and it seemed to work.

![image](https://github.com/dotnet/sdk/assets/23152278/2eaf8bca-75ae-4d64-9747-2889f7123988)

dsplaisted commented:
> Doing the testing now. One thing I noticed is these messages side by side might confuse people but I think it's ok, since the 2nd message is prefixed with 'VS' workloads.

I think we should update this so it doesn't print the "no workloads installed" message if there are any VS workloads installed.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/list/VisualStudioWorkloads.cs:
```diff
@@ -119,6 +122,36 @@ internal static class VisualStudioWorkloads
+            IEnumerable<WorkloadId> workloadsWithExistingInstallRecords, IReporter reporter)
+        {
+            // Do this check to avoid adding an unused & unnecessary method to FileBasedInstallers
+            if (OperatingSystem.IsWindows() && typeof(NetSdkMsiInstallerClient) == workloadInstaller.GetType())
```
Is there a reason we couldn't call `WorkloadInstallType.GetWorkloadInstallType()` here instead of doing a .NET type comparison?

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/list/VisualStudioWorkloads.cs:
```diff
@@ -119,6 +122,36 @@ internal static class VisualStudioWorkloads
+                {
+                    reporter.WriteLine(
+                        string.Format(LocalizableStrings.WriteCLIRecordForVisualStudioWorkloadMessage,
+                        string.Join(", ", workloadsToWriteRecordsFor.Select(w => w.ToString()).ToArray()))
```
I think separating lists with a comma isn't well localized, although we probably do it in various places.  In some cases we put each element on a separate line.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/list/VisualStudioWorkloads.cs:
```diff
@@ -119,6 +122,36 @@ internal static class VisualStudioWorkloads
+                    );
+                }
+
+                ((NetSdkMsiInstallerClient)workloadInstaller).WriteWorkloadInstallRecords(workloadsToWriteRecordsFor);
```
I don't think that we necessarily need to add a new method to `NetSdkMsiInstallerClient` and do an explicit cast here.  I think instead we could do something like:

```c#
foreach (var workload in workloadsToWriteRecordsFor)
{
    workloadInstaller.GetWorkloadInstallationRecordRepository().WriteWorkloadInstallationRecord(workload, new SdkFeatureBand(workloadResolver.GetSdkFeatureBand()));
}
```

We'd lose the additional log message, but that doesn't seem critical.

dsplaisted reviewed (commented):
Looks good.  I had a few comments, but nothing that necessarily needs to be addressed.
